qiskit/transpiler/passes/routing/bip_mapping.py

```diff
@@ -0,0 +1,225 @@
+
+        if len(dag.qubits) != self.coupling_map.size():
+            raise TranspilerError(
+                "BIPMapping pass requires the number of virtual and physical qubits are the same."
+            )
+
+        if self.property_set["layout"]:
+            logger.info("BIPMapping ignores any given layout.")
+
+        original_dag = dag
+
+        dummy_steps = math.ceil(math.sqrt(dag.num_qubits()))
+        if self.max_swaps_inbetween_layers is not None:
+            dummy_steps = max(0, self.max_swaps_inbetween_layers - 1)
+
+        model = BIPMappingModel(
+            dag=dag, coupling_map=self.coupling_map, dummy_timesteps=dummy_steps
+        )
+
+        if len(model.su4layers) == 0:
+            logger.info("BIPMapping is skipped due to no 2q-gates.")
+            return original_dag
+
+        model.create_cpx_problem(objective=self.objective)
+
+        try:
+            model.solve_cpx_problem(time_limit=self.time_limit, threads=self.threads)
+        except TranspilerError as err:
+            logger.warning("%s dag was not mapped in BIPMapping.", err.message)
+            return original_dag
+
+        # Get the optimized initial layout
+        optimized_layout = model.get_layout(0)
+
+        # Create a layout to track changes in layout for each layer
+        layout = copy.deepcopy(optimized_layout)
+
+        # Construct the mapped circuit
+        canonical_register = QuantumRegister(self.coupling_map.size(), "q")
+        mapped_dag = self._create_empty_dagcircuit(dag, canonical_register)
+        interval = dummy_steps + 1
+        for k, layer in enumerate(dag.layers()):
+            if model.is_su4layer(k):
+                su4dep = model.to_su4layer_depth(k)
+                # add swaps between (su4dep-1)-th and su4dep-th su4layer
+                from_steps = max(interval * (su4dep - 1), 0)
+                to_steps = min(interval * su4dep, model.depth - 1)
+                for t in range(from_steps, to_steps):  # pylint: disable=invalid-name
+                    for (i, j) in model.get_swaps(t):
+                        mapped_dag.apply_operation_back(
+                            op=SwapGate(),
+                            qargs=[canonical_register[i], canonical_register[j]],
+                        )
+                        # update layout, swapping physical qubits (i, j)
+                        layout.swap(i, j)
+
+            # map gates in k-th layer
+            for node in layer["graph"].nodes():
+                if node.type == "op":
+                    mapped_dag.apply_operation_back(
+                        op=copy.deepcopy(node.op),
+                        qargs=[canonical_register[layout[q]] for q in node.qargs],
+                        cargs=node.cargs,
+                    )
+                # TODO: double check with y values?
+
+        # Check final layout
+        final_layout = model.get_layout(model.depth - 1)
+        assert layout == final_layout
```

Comment:
Qiskit tries to avoid assertion checks. The preferred way is something along these lines:
```suggestion
        if layout != final_layout:
             raise TranspilerError(....)
```